src/messageCache.js

```diff
@@ -0,0 +1,115 @@
+const RPC = require('./message').rpc.RPC
+const utils = require('./utils')
+/**
+ * This file implements the Message Cache API provided in https://github.com/libp2p/go-libp2p-pubsub/blob/master/mcache.go#L15 used in gossip sub to store messages that were sent for the last few heartbeat ticks.
+ */
+
+
+class CacheEntry {
+
+    /**
+     * @param {String}
+     * @param {String[]}
+     * @constructor
+     */
+    constructor (msgID, topics) {
+        this.msgID = msgID
+	this.topics = topics
+    }
+
+}
+
+class MessageCache {
+
+    /**
+     * @param {Number} gossip
+     * @param {Number} history
+     *
+     * @constructor
+     */
+    constructor (gossip, history) {
+	/**
+	 * @type {Map<string, RPC.Message>}
+	 */
+        this.msgs = new Map()
+
+	/**
+	 * @type {Array<Array{CacheEntry}>}
+	 */
+	this.history = []
```

Comment:
Oddly, making this change breaks some of the tests I have written. I have verified that both methods for populating the multidimensional `this.history` array are equivalent. I have tried searching for why this might be but found conflicting reasons. I will keep it as it is for the moment.